trin-core/src/types/accumulator.rs

```diff
@@ -0,0 +1,219 @@
+        let header_record = HeaderRecord {
+            block_hash: new_block_header.hash(),
+            total_difficulty: last_total_difficulty + new_block_header.difficulty,
+        };
+        self.current_epoch
+            .header_records
+            .push(header_record)
+            .expect("Invalid accumulator state, more current epochs than allowed.");
+    }
+}
+
+/// Datatype responsible for storing all of the merkle roots
+/// for epoch accumulators preceding the current epoch.
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct HistoricalEpochs {
+    epochs: HistoricalEpochsList,
+}
+
+impl TreeHash for HistoricalEpochs {
+    fn tree_hash_type() -> tree_hash::TreeHashType {
+        tree_hash::TreeHashType::List
+    }
+
+    fn tree_hash_packed_encoding(&self) -> Vec<u8> {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_packing_factor() -> usize {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_root(&self) -> tree_hash::Hash256 {
+        // list of composite objects
+        // mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))
+        let hash_tree_roots: Vec<tree_hash::Hash256> = self
+            .epochs
+            .iter()
+            .map(|epoch_root| epoch_root.tree_hash_root())
+            .collect();
+        let mut historical_epochs_hasher = MerkleHasher::with_leaves(MAX_HISTORICAL_EPOCHS);
+        for root in &hash_tree_roots {
+            historical_epochs_hasher
+                .write(root.as_bytes())
+                .expect("Invalid master accumulator state: Too many epochs.");
+        }
+        let historical_epochs_root = historical_epochs_hasher.finish().unwrap();
+        tree_hash::mix_in_length(&historical_epochs_root, hash_tree_roots.len())
+    }
+}
+
+/// Data type responsible for maintaining a store
+/// of all header records within an epoch.
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct EpochAccumulator {
+    header_records: HeaderRecordList,
+}
+
+impl TreeHash for EpochAccumulator {
+    fn tree_hash_type() -> tree_hash::TreeHashType {
+        tree_hash::TreeHashType::List
+    }
+
+    fn tree_hash_packed_encoding(&self) -> Vec<u8> {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_packing_factor() -> usize {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_root(&self) -> tree_hash::Hash256 {
+        // list of composite objects
+        // mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))
```

Comment:
Is this comment leftover from some previous implementation?

Reply:
Can we remove?

Reply:
Same reply as above.